### ISSUE TITLE
GeoTIFF v2 proposal: embed WKT

### ISSUE THREAD

**Issue by rouault** [opengeospatial/geotiff#56]
Of course not to be considered for the current work, but perhaps worth considering for a GeoTIFF v2.0:

A lot of the issues we have to face with the myriad of GeoTIFF keys could be solved easily by having one single GeoKey which would be a ASCII GeoKey "WKTCitation" and just rely on the wkt-crs standard

**Comment by rouault** [comment]
Note: there is a prior art in GeoTIFF regarding using WKT in ESRI implementations where PCSCitationGeoKey might have as value "ESRI PE String = <esri_wkt_string>"

**Comment by schwehr** [comment]
Support for GOES would be helpful:

As noted here: https://lists.osgeo.org/pipermail/proj/2019-June/008675.html

```
gsutil cp
gs://gcp-public-data-goes-17/ABI-L1b-RadC/2019/151/21/OR_ABI-L1b-RadC-*_G17_s20191512131197_*.nc .

gdalinfo NETCDF:OR_ABI-L1b-RadC-M6C02_G17_s20191512131197_e20191512133570_c20191512133595.nc:Rad
Driver: netCDF/Network Common Data Format
Files: OR_ABI-L1b-RadC-M6C02_G17_s20191512131197_e20191512133570_c20191512133595.nc
Size is 10000, 6000
Coordinate System is:
PROJCS["unnamed",
    GEOGCS["unknown",
        DATUM["unknown",
            SPHEROID["Spheroid",6378137,298.2572221]],
        PRIMEM["Greenwich",0],
        UNIT["degree",0.0174532925199433]],
    PROJECTION["Geostationary_Satellite"],
    PARAMETER["central_meridian",-137],
    PARAMETER["satellite_height",35786023],
    PARAMETER["false_easting",0],
    PARAMETER["false_northing",0],
    EXTENSION["PROJ4","+proj=geos +lon_0=-137 +h=35786023 +x_0=0 +y_0=0 +ellps=GRS80 +units=m +no_defs  +sweep=x"]]
Origin = (-2505021.493779813405126,1583173.639055640902370)
Pixel Size = (501.004328871885207,501.004328871885264)
Metadata:
  goes_imager_projection#grid_mapping_name=geostationary
  goes_imager_projection#inverse_flattening=298.2572221
  goes_imager_projection#latitude_of_projection_origin=0
  goes_imager_projection#longitude_of_projection_origin=-137
  goes_imager_projection#long_name=GOES-R ABI fixed grid projection
  goes_imager_projection#perspective_point_height=35786023
  goes_imager_projection#semi_major_axis=6378137
  goes_imager_projection#semi_minor_axis=6356752.31414
  goes_imager_projection#sweep_angle_axis=x
[SNIP]

gdalinfo --version
GDAL 2.2.4, released 2018/03/19
```

**Comment by joanma747** [comment]
I would like to emphasize that a the georeference of a geostacionary satellite image cannot be express with the current keys and there are (at least) two solutions:
* Adopt WKT
* Extend the geokeys with the necessary parameters

**Comment by gerritholl** [comment]
Isn't geostationary projection already supported at http://geotiff.maptools.org/proj_list/geos.html ?  Am I missing something?

> I would like to emphasize that a the georeference of a geostacionary satellite image cannot be express with the current keys and there are (at least) two solutions:

Isn't geostationary projection already supported at http://geotiff.maptools.org/proj_list/geos.html ?  Am I missing something?

http://geotiff.maptools.org/proj_list/geos.html

```
$ gdalinfo myfile.tif
Driver: GTiff/GeoTIFF
Files: myfile.tif
Size is 20000, 20000
Coordinate System is:
PROJCRS["WGS 84 / World Equidistant Cylindrical",
    BASEGEOGCRS["WGS 84",
        ENSEMBLE["World Geodetic System 1984 ensemble",
            MEMBER["World Geodetic System 1984 (Transit)"],
            MEMBER["World Geodetic System 1984 (G730)"],
            MEMBER["World Geodetic System 1984 (G873)"],
            MEMBER["World Geodetic System 1984 (G1150)"],
            MEMBER["World Geodetic System 1984 (G1674)"],
            MEMBER["World Geodetic System 1984 (G1762)"],
            MEMBER["World Geodetic System 1984 (G2139)"],
            ELLIPSOID["WGS 84",6378137,298.257223563,
                LENGTHUNIT["metre",1]],
            ENSEMBLEACCURACY[2.0]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4326]],
    CONVERSION["World Equidistant Cylindrical",
        METHOD["Equidistant Cylindrical",
            ID["EPSG",1028]],
        PARAMETER["Latitude of 1st standard parallel",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8823]],
        PARAMETER["Longitude of natural origin",0,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["False easting",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",0,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["easting (X)",east,
            ORDER[1],
            LENGTHUNIT["metre",1]],
        AXIS["northing (Y)",north,
            ORDER[2],
            LENGTHUNIT["metre",1]],
    USAGE[
        SCOPE["Graticule coordinates expressed in simple Cartesian form."],
        AREA["World."],
        BBOX[-90,-180,90,180]],
    ID["EPSG",4087]]
Data axis to CRS axis mapping: 1,2
Origin = (-10016827.231008989736438,10016827.231008989736438)
Pixel Size = (1001.682778760644396,-1001.682723100899011)
Metadata:
   ...
```

**Comment by jerstlouis** [comment]
The current situation to figure out a CRS from a myriad of GeoTIFF keys is a real nightmare.
Embedding an AS Topic 2 CRS encoding (whether CRS/PROJ-JSON or WKT-CRS) is the sane solution.
Should be at the top of the list for 2.0! :)

**Comment by joanma747** [comment]
Does this imply removing support for the GeoKeys that result redundant?. It could be good to have a clean start with no conflicting ways to express the same.

**Comment by desruisseaux** [comment]
Note that GeoKeys are already redundant with EPSG codes in the GeoTIFF 1 specification. Introducing WKT in GeoTIFF would not really change the redundancy situation. This redundancy can be a useful fallback when the GeoTIFF is read with a software having no connection to the EPSG database and no capability to parse WKT. In geospatial communities (OGC, OSGeo…), using a geospatial library such as GDAL or Apache SIS may seem an evidence. But in other communities (Apache, big-data, etc.), developers sometime want to stick to the non-geospatial library they are used to.

**Comment by rouault** [comment]
> But in other communities (Apache, big-data, etc.), developers sometime want to stick to the non-geospatial library they are used to.

I'm not sure how you can make sense of the geokeys of GeoTIFF 1.0 or 1.1 that would be obsoleted by a WKT/JSON CRS encoding, without serious CRS understanding capabilities. The only capability which doesn't require too advance knowledge I can think of, and that can be done relatively easily with GeoTIFF 1.x, is perhaps to extract the CRS EPSG code, and hand it over to something else. That would actually be an argument in favoring a JSON CRS encoding, which would make it easy to extract it. cf the hints at https://github.com/opengeospatial/geoparquet/blob/main/format-specs/geoparquet.md#ogccrs84-details aimed at non super CRS savy users to extract the CRS code. But figuring out datum, projection method, projection parameters already require a pretty advanced knowledge of the CRS business.
IMHO having a GeoTIFF 2.0 keeping all the geokeys defining datum, units, projection method&parameters, etc + allowing WKT/JSON would be the worse outcome, since it would require readers to know both worlds.

**Comment by hobu** [comment]
> IMHO having a GeoTIFF 2.0 keeping all the geokeys defining datum, units, projection method&parameters, etc + allowing WKT/JSON would be the worse outcome, since it would require readers to know both worlds.

💯 concur. Especially because it is not always possible to express equivalent definitions and you inevitably end up with downstream consumers having to decide which definition is the "most correct" given their circumstances.

**Comment by rouault** [comment]
Coming from a remark posted at https://lists.osgeo.org/pipermail/proj/2024-April/011351.html, if we allow WKT or JSON CRS encoding, we should be careful about the text encoding. Current GeoTIFF keys are limited to ASCII, whereas WKT allows for Latin-1 character set (encoded as UTF-8) for thing like CRS names, and possibly full Unicode for other auxiliary metadata. So a GeoTIFF key holding a WKT or JSON CRS should support UTF-8

**Comment by rouault** [comment]
https://lists.osgeo.org/pipermail/proj/2024-April/011353.html: """let me 
point out that the root of the problem is the TIFF6.0 standard that 
continues to define field type 2 as ASCII. Their exact wording is: 
"8-bit byte that contains a 7-bit ASCII code; the last byte must be NUL 
(binary zero)" (page 15)

IMO everything would work just fine if this would be replaced with 
"8-bit byte that contains UTF8 code units; the last byte must be NUL 
(binary zero)". Given that UTF-8 is an extension of ASCII, backwards 
compatibility wouldn't be so bad.
"""

**Comment by rouault** [comment]
@joanma747 Wasn't there a discussion some time ago about having a "poll" to GeoTIFF users to know if they'd be willing to have a breaking GeoTIFF 2.0 ?

**Comment by jjimenezshaw** [comment]
In my wish list for GeoTIFF 2.0 I would add "always bigtiff". We are in the 64 bit era.

**Comment by joanma747** [comment]
There where some discussions and some consensus on having a new way to specify the georeference based on embedding WKT and JSON encodings for CRS that will definitely break compatibility with the previous versions. Having bigtiff as an standard and incorporating into GeoTIFF was in our radar too.
As a chair of the group, it is my fault (not only mine) to keep delaying all this due to lack of time. Actually the Testbed 20 activities on HEIF format and the possible standard for a GeoHEIF has broad new ideas that can be also applied to the GeoTIFF.

**Comment by rprinceley** [comment]
We (Esri) are also seeking an official method for embedding WKT. We have considered expanding our use of PCSCitationGeoKey in ArcGIS, but we would prefer an official solution.

**Comment by joanma747** [comment]
Great to hear about it. Lets see if we can align the work of GeoHEIF with GeoTIFF and make some progress in this respect.

**Comment by rouault** [comment]
> Lets see if we can align the work of GeoHEIF with GeoTIFF

I'm afraid this will lead to an over-engineered solution. Looking at https://docs.ogc.org/per/24-038r1.html#req-mcrs, it allows 4 different flavors: WKT2, CURIE, CRS URL and CRS JSON. That's 3 more than what we need.

(That said, switching to WKT2 wouldn't fully solve the interoperability issue for the Geostationnary satellite projection as there's no normalized way of encoding it to WKT (PROJ made its own choice). Maybe starting from https://github.com/cf-convention/cf-conventions/blob/main/appf.adoc#geostationary-projection could be a good idea. But that's a bit of an orthogonal issue)

**Comment by jerstlouis** [comment]
@rouault I would personally prefer if we had at least two options:
- for referencing (no need to parse -- recognize existing CRS), or
- embedded definition (needs parser, for custom cases for which no registration exists).

For the referencing, I have a preference for CURIEs which avoids the version / http vs https / lowercase vs. uppercase authority.

I would also much prefer PROJJSON or CRS-JSON over WKT for the embedded definitions.

**Comment by rouault** [comment]
> I would personally prefer if we had at least two options:

if we think in terms of how a GeoTIFF 2.0 would be deployed/implemented, as this would be a major breaking change, taking probably a decade to be widely supported in the industry, I would anticipate that writers would still write GeoTIFF 1.0/1.1 for any EPSG CRS, and only use GeoTIFF 2.0 when GeoTIFF 1.x cannot be used (unsupported projection methods, supported projection method but non EPSG CRS where it is important to have the autority & code preserved). At least that's what GDAL would certainly do. So the need for a GeoTIFF 2.0 supporting codes/CURIEs, without requiring a full CRS definition, doesn't seem completely obvious to me. A limitation of the code approach is that as new codes keep being added (think to the new NSRS2022 codes that are going to be added this year), you need access to a (up-to-date) CRS registry, query it and make sense of the result. The difficulty of that or having a whatever-text-CRS-encoding parser are quite comparable.
If you look at GeoPackage for example, it is a bit in the middle. You *have* to write a WKT in the gpkg_spatial_ref_sys table AND (organization, organization_coordsys_id), the later being necessarily set to dummy/non-interoperable values when writing a custom WKT.

And yes, a JSON encoding could be more friendly to readers (and also simplifying the access to the CRS code potentially embedded in the definition). CRS-JSON having been at a unpublished draft stage for several years, it would postpone GeoTIFF 2.0 by several years.

**Comment by jerstlouis** [comment]
> I would anticipate that writers would still write GeoTIFF 1.0/1.1 for any EPSG CRS, and only use GeoTIFF 2.0 when GeoTIFF 1.x cannot be used

I have hope for other improvements to a GeoTIFF 2.0 worth the switch, such as field descriptions (like newly published [Common - Part 3: Schemas](https://docs.ogc.org/is/23-058r2/23-058r2.html)), more than 2 real dimensions (as in [Common-2 Uniform Additional Dimensions](https://docs.ogc.org/DRAFTS/20-024.html#_uml_class_model)), per field data type and per-field resolution (e.g., for higher-res panchromatic bands). Not sure to what extent TIFF can support the latter.

**Comment by rouault** [comment]
> such as field descriptions (like newly published [Common - Part 3: Schemas](https://docs.ogc.org/is/23-058r2/23-058r2.html))

Band description in GDAL_METADATA

> more than 2 real dimensions (as in [Common-2 Uniform Additional Dimensions](https://docs.ogc.org/DRAFTS/20-024.html#_uml_class_model))

you mean 3D or 4D arrays ? That's no longer TIFF ==> use netCDF, HDF5, Zarr, etc.

> per field data type

In theory TIFF supports per-band/channel data type and bit width. In practice libtiff implementation doesn't (and probably not many alternative implementations)

> per-field resolution

all band/channel in a same TIFF IFD must have the same size. You can potentially have several IFDs of different size in a same TIFF file, potentially each with their own GeoTIFF keys. Moderately well supported

Let's not make TIFF what it is isn't. There are plenty of other formats for that.

**Comment by jerstlouis** [comment]
> Band description in GDAL_METADATA

That could potentially be the mechanism that gets standardized, along with `GDAL_NODATA`.

> you mean 3D or 4D arrays ? That's no longer TIFF ==> use netCDF, HDF5, Zarr, etc.

Yes, n-D arrays. But it's relatively simple to support this by stacking TIFF images, with the proper metadata that would distinguish between dimensions (domain) and fields (range).

> Let's not make TIFF what it is isn't. There are plenty of other formats for that.

TIFF has a nice simplicity / low dependency to it, and it would be great if it could also support the multidimensional  use cases.
HDF5 and netCDF 4 depending on it in particular is a dependency nightmare.
Zarr is mainly about chunking, but if you're focusing on a specific chunk (as for tiles), this is not useful.

**Comment by rouault** [comment]
> TIFF has a nice simplicity / low dependency to it, and it would be great if it could also support the multidimensional use cases.

I've tried that few years ago in Testbed 17 and wasn't super convinced by the result: https://docs.ogc.org/per/21-032.html#toc14 (to be fully honest I reluctantly did the experiment being unconvinced even before starting it). IMHO this is stretching TIFF beyond what it has been designed for. A native n-D support in TIFF *if needed* should come with a distinct name.  Or look at [OME-TIFF](https://ome-model.readthedocs.io/en/stable/ome-tiff/specification.html) but from a quick look this is conceptually similar from my testbed experiment, relying on multi-IFD to store dimensions > 2 and some metadata to explain how to reconstruct n-D from 2D planes. Note that OME-TIFF is ported to Zarr as OME-ZARR: https://link.springer.com/article/10.1007/s00418-023-02209-1
If you create a .tar file (or uncompressed .zip) with the zarr.json and a sharded array with the shard size = array size, then you get a single 0.0.0... data file, which overall gives you a single file with pretty similar characteristics to a COG, except it was designed for n-D.

**Comment by jerstlouis** [comment]
@rouault Thanks for pointing out OME-TIFF.

The Testbed 17 limitations seem mostly about very large single-file COG datasets.
I get that this is not ideal, and that the sequential IFD lookup will impact performance.

The primary use case I'm interested in is for nD tiles (where each multi-dimensional is stored as separate files, providing the "chunking") and OGC API - Coverage responses (with subsetting and/or downscaling), where the total file size would be relatively small (from a few kilobytes up to a gigabyte or so), and this should not be a significant issue.

**Comment by mdsumner** [comment]
> Yes, n-D arrays. But it's relatively simple to support this by stacking TIFF images, with the proper metadata that would distinguish between dimensions (domain) and fields (range).

 it is easy, but please leave that downstream hacking for languages where it already exists in abundance and for which STAC provides an already sophisticated improvement.   (The Zarr world is busily reading IFD tables with Rust in order to stack "virtual chunks" into ever more complicated array trees, I'd suggest adopt that or as Even implied 'just say no'). And, fwiw GDAL multidim VRT already provides a better solution.

**Comment by rprinceley** [comment]
> if we think in terms of how a GeoTIFF 2.0 would be deployed/implemented, as this would be a major breaking change, taking probably a decade to be widely supported in the industry, I would anticipate that writers would still write GeoTIFF 1.0/1.1 for any EPSG CRS, and only use GeoTIFF 2.0 when GeoTIFF 1.x cannot be used (unsupported projection methods, supported projection method but non EPSG CRS where it is important to have the autority & code preserved). At least that's what GDAL would certainly do.

This is similar to what we already do in ArcGIS. We prioritize PE WKT to support non-EPSG (Esri) CRS, write GeoKeys when possible, otherwise set keys to custom, and include citation strings. This works well for us, but other parsers (except GDAL) don't support PE WKT.

Since v2 will be an add-on and we will continue to write 1.x keys, we should proceed with this.

**Comment by desruisseaux** [comment]
Regarding CRS-JSON, the specification draft started as [OGC 26-009](https://docs.ogc.org/DRAFTS/26-009.html) using the WKT 2 specification as a template for the document structure. The [JSON schema](https://github.com/opengeospatial/CRS-JSON-Encoding/blob/main/uml2json/schemas/iso19111.json) has been initially generated with ShapeChange. [A prototype](https://github.com/Geomatys/CRS-JSON-Encoding/tree/main/prototype) brings CRS-JSON support in Apache SIS, GeoTools, PROJ4J and PROJ-JNI, and a [demonstration server](https://demo.geomatys.com/crs-json/index.html) is online.

CRS-JSON depends strongly on [ISO 19115-4](https://www.iso.org/standard/86968.html), which entered in approval phase 2 weeks ago. But the JSON schema [can already be donwloaded](https://schemas.isotc211.org/19115/-4/mdj/1.0.0/json/browse/#mdj) — note that this schema has also been generated by ShapeChange. An example is [available here](https://github.com/ISO-TC211/XML/blob/master/schemas.isotc211.org/json/19115/-4/C1-19115-4-JSON-example.json) (click on the "Raw" button for viewing the actual JSON file).

Integration with ISO 19115-4 is the point of CRS-JSON. For example, above-cited example is pretty cool, and it would be nice to have the ISO 19115 bounding box of a CRS viewable in such demo. However, the ShapeChange output of the ISO 19115 schema is slightly different than the ShapeChange output of CRS-JSON, maybe because the latter was generated 20 months ago. In particular, `entityType` seems to have been renamed as `type`. Which is why we needed to see what ISO 19115-4 will look like before to finalize CRS-JSON.
